common/events_io.hpp

```diff
@@ -124,6 +124,21 @@ class AsyncWriterFromWriter : virtual public mio::AsyncWriter {
 	EventLoop &loop_;
 };
 
+using AsyncReaderFromEventLoopFunc = function<mio::AsyncReaderPtr(EventLoop &loop)>;
+
+class ReaderFromAsyncReader : virtual public mio::Reader {
+public:
+	// Since the async reader usually needs the event loop object, we need to use this factory
+	// function to create the async reader inside of the constructor.
+	ReaderFromAsyncReader(AsyncReaderFromEventLoopFunc func);
+
+	mio::ExpectedSize Read(vector<uint8_t>::iterator start, vector<uint8_t>::iterator end) override;
+
+private:
+	EventLoop event_loop_;
```

Comment:
Where does `event_loop_` come from? I don't see any constructor here and the factory function seems to be relying on it.

Reply:
It's just a default-constructed `EventLoop`, which is independent of the main application event loop. This is specifically to make this `Reader` do only one thing: Read from the `AsyncReader`.

But this is indeed quite unobvious. I'll document it next to the `event_loop_` variable.